step4 구현 - ㅇㅁㅇ!!!!!!

isisrin opened the pull request:
- DeleteHistoryGenerator 추가해봄
 - DeleteHistory Builder 추가
 - Repository에 findByContentId 추가 (queryDsl, specification 관련 피드백 하지 말아주세요ㅠㅠ)

isisrin committed:
step4 구현 - ㅇㅁㅇ!!!!!!
 - DeleteHistoryGenerator 추가해봄
 - DeleteHistory Builder 추가
 - Repository에 findByContentId 추가 (queryDsl, specification피드백 하지 말아주세요ㅠㅠ)
 -

jojoldu commented:
@isisrin 네 ㅋㅋㅋㅋㅋ querydsl 피드백은 제외하겠습니다 ㅋㅋ

jojoldu commented on src/test/java/nextstep/service/QnaServiceTest.java:
```diff
@@ -28,43 +28,57 @@
 
+    @Autowired
+    private DeleteHistoryRepository deleteHistoryRepository;
+
```
조금 더 좋은 테스트 코드로 가볼까요?

만드신 테스트들에는 한가지 문제가 있는데요.
테스트가 끝나고 저장/삭제/수정된 데이터들이 계속 남아 있습니다
단일 테스트를 수행시엔 큰 문제가 안되는데요.
전체 테스트를 수행시엔 앞에서 수행되 저장된 데이터들이 남아 있어, 이후 테스트들이 영향을 받을 수 있습니다.

아래와 같이 전체 테스트를 한번 수행해보시고

![image](https://user-images.githubusercontent.com/7760496/55725944-ab073700-5a49-11e9-9168-cbf877e45d36.png)

@after 메소드를 만들어 사용된 Repository는 초기화 하시는것을 추천드려요

repository.deleteAll()를 사용하면 됩니다 :)

jojoldu commented on src/main/java/nextstep/domain/DeleteHIstoryGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package nextstep.domain;
+
+public interface DeleteHIstoryGenerator {
```
I가 대문자입니다
소문자로 변경해주세요.

jojoldu reviewed (changes requested):
@isisrin 혜린님 코드가 굉장히 좋아졌습니다.
조금만 더 하시면 될것 같습니다 :)

하신것 중에 스프링부트 2가 되면서 발생한 문제와 **전체 테스트 수행시 문제가 되는** 등의 이슈에 대해 남겼으니 수정 해주시면 됩니다 :)

isisrin committed:
step4 안 올라간 것 추가.....

isisrin committed:
step4 피드백반영
 - 안쓰는 Fixture삭제
 - DeleteHistoryGenerator 오타수정 (매의눈 이시네요!!)
 - (strategy = GenerationType.IDENTITY) 추가
 - findByContentId Optional로 바꿈!
 - @EqualsAndHashCode 삭제

isisrin committed:
step4 피드백반영
 - @EqualsAndHashCode 삭제
 - Optinal추가
 - TestCode 실제로 repository 이용하도록 수정

jojoldu commented on src/main/java/nextstep/domain/DeleteHistory.java:
```diff
@@ -10,7 +10,7 @@
 public class DeleteHistory {
     @Id
-    @GeneratedValue
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
피드백 반영 좋습니다 :)

jojoldu commented on src/main/java/nextstep/domain/DeleteHistoryRepository.java:
```diff
@@ -2,8 +2,9 @@
+import java.util.Optional;
 
-    DeleteHistory findByContentId(Long contentId);
+public interface DeleteHistoryRepository extends CrudRepository<DeleteHistory, Long> {
```
CrudRepository 보다는 JpaRepository를 상속받는게 좋습니다
좀 더 확장된 기능을 제공하기도 하고, **레퍼런스 문서에서도 예제로 사용됩니다**

jojoldu commented on src/main/java/nextstep/domain/Question.java:
```diff
@@ -87,12 +88,32 @@ public Question update(User loginUser, Question question) throws UnAuthenticatio
 
+    private void checkAnswerByWriter() throws CannotDeleteException {
+        Optional<Answer> filteredAnser = answers.stream().filter(answer -> answer.isNotOwner(writer)).findAny();
+        if(filteredAnser.isPresent()) {
```
이렇게 구현하시기 보다는 

```java
long notOwnerCount = answers.stream().filter(answer -> answer.isNotOwner(writer)).count();
if(notOwnerCount > 0) {
throw new CannotDeleteException("다른 사람의 답변까지 지울 수 없어요!");
} 
```

가 좀 더 의미가 명확하게 전달될것 같습니다 :)

jojoldu commented on src/main/java/nextstep/service/DeleteHistoryService.java:
```diff
@@ -26,7 +27,7 @@ public DeleteHistory save(DeleteHistory deleteHistory) {
 
     public DeleteHistory findByContentId(Long contentId) {
-        return deleteHistoryRepository.findByContentId(contentId);
+        return deleteHistoryRepository.findByContentId(contentId).orElseThrow(EntityNotFoundException::new);
```
피드백 반영 💯

jojoldu commented on src/test/java/nextstep/service/QnaServiceTest.java:
```diff
@@ -102,25 +130,39 @@ public void setUp() {
         qnaService.deleteAnswer(MOCK_USER, 0);
     }
+
+    @After
```
after는 before 바로 다음에 두시는게 이후에 알아보기 쉬울것 같습니다 :)

jojoldu commented on src/test/java/nextstep/web/api/ApiQnaAcceptanceTest.java:
```diff
@@ -67,13 +73,27 @@
 
     @Test
-    public void 질문_삭제() {
+    public void 질문_삭제_성공_테스트_답변0개() {
```
메소드명이 좋습니다 :)
조금 더 명확한 이름이 되려면

```java
질문_삭제_성공시_답변은_0개가_된다()
```
같은 식으로 지으셔도 좋습니다.
굳이 메소드명에 테스트가 들어가기 보다는 **어떤 상황에서 어떤 결과가 나와야하는지**를 메소드명으로 해주셔도 좋을것 같아요 :)

jojoldu reviewed (changes requested):
@isisrin 혜린님 정말 코드가 좋아졌습니다!
고생하셨습니다.
이번엔 마지막 피드백입니다 :)
반영해주시면 최종 Merge 하겠습니다!

isisrin committed:
step4 피드백반영
 - JpaRepository로 바꿈
 - findAny -> count로 변경
 - @After위치 바꿈

jojoldu commented on src/main/java/nextstep/domain/Question.java:
```diff
@@ -97,8 +97,8 @@ public void deleteQuestion() throws CannotDeleteException { //throws CannotDelet
-        Optional<Answer> filteredAnser = answers.stream().filter(answer -> answer.isNotOwner(writer)).findAny();
-        if(filteredAnser.isPresent()) {
+        long notOwnerCount = answers.stream().filter(answer -> answer.isNotOwner(writer)).count();
+        if(notOwnerCount > 0) {
```
피드백 반영 💯

jojoldu reviewed (approved):
@isisrin 피드백 반영 모두 확인했습니다 :)
고생하셨습니다!
해당 Step4 완료하겠습니다!
